### PR TITLE
Update arkworks to 0.4.0

### PR DESCRIPTION
**Pull request by imikushin**
Implements #980

### CODE REVIEW

**Comment by mimoo** [comment]
Some notes to self (and other interested people):

* to get this in, we need a matching PR on the Mina side
* some other members are working on other changes at the moment, we should compile a list of the changes to be aware of from 0.3.0 to 0.4.0 so that they can easily handle merge conflicts

Hopefully we can get this in sooner than later, as this is a great PR!

**Comment by imikushin** [comment]
Looks like the CI got broken by some stuff that is now in `master`.
cc @mimoo @mrmr1993

**Comment by imikushin** [comment]
Yep. See https://github.com/o1-labs/proof-systems/actions/runs/4218246874/jobs/7322611037 fail after commit https://github.com/o1-labs/proof-systems/pull/1021/commits/1d64e966d5cff2e61f98f603743fb67c8472d882 in #1021

The commit adds a line to CHANGELOG.md which (hopefully) doesn't have anything to do with
```
error: failed to run custom build command for `ocaml-sys v0.22.3`
```
which is also how the CI is broken for this PR.

**Comment by imikushin** [comment]
#1021 needs to be merged first.

**Comment by imikushin** [comment]
@mimoo, to your comment https://github.com/o1-labs/proof-systems/pull/998#issuecomment-1435941384:

Here's a [short guide to relevant changes](https://hackmd.io/@imikushin/B1rjCZeAs) in arkworks.

There's also official [CHANGELOG.md](https://github.com/arkworks-rs/algebra/blob/master/CHANGELOG.md)

**Comment by imikushin** [comment]
Also, I believe, this is now ready to merge.

File: poly-commitment/src/combine.rs [hunk 1]
```diff
@@ -585,127 +586,122 @@ pub fn affine_shamir_window_table_one<P: SWModelParameters>(
...
-            ((false, false), (true, false)) => res.add_assign_mixed(&g00_10),
-            ((false, true), (true, false)) => res.add_assign_mixed(&g01_10),
-            ((true, false), (true, false)) => res.add_assign_mixed(&g10_10),
-            ((true, true), (true, false)) => res.add_assign_mixed(&g11_10),
-
-            ((false, false), (true, true)) => res.add_assign_mixed(&g00_11),
-            ((false, true), (true, true)) => res.add_assign_mixed(&g01_11),
-            ((true, false), (true, true)) => res.add_assign_mixed(&g10_11),
-            ((true, true), (true, true)) => res.add_assign_mixed(&g11_11),
+            ((false, true), (false, false)) => res.add_assign(&g01_00),
```
Comment (mimoo) [inline poly-commitment/src/combine.rs:603]: `res += &g01_00` would be nicer here : o

File: poly-commitment/src/commitment.rs [hunk 2]
```diff
@@ -231,7 +236,7 @@ impl<C: AffineCurve> PolyComm<C> {
             return Self::new(vec![C::zero()], None);
         }
 
-        let all_scalars: Vec<_> = elm.iter().map(|s| s.into_repr()).collect();
+        let all_scalars: Vec<_> = elm.to_vec();
```
Comment (mimoo) [inline poly-commitment/src/commitment.rs:239]: we could delete this line now

File: poly-commitment/src/commitment.rs [hunk 3]
```diff
@@ -574,15 +574,15 @@ impl<G: CommitmentCurve> SRS<G> {
         let basis_len = self.g.len();
         let coeffs_len = plnm.coeffs.len();
 
-        let coeffs: Vec<_> = plnm.iter().map(|c| c.into_repr()).collect();
+        // let coeffs: Vec<_> = plnm.iter().map(|c| c.into_bigint()).collect();
```
Comment (mimoo) [inline poly-commitment/src/commitment.rs:577]: delete

File: poly-commitment/src/commitment.rs [hunk 4]
```diff
@@ -866,8 +866,7 @@ impl<G: CommitmentCurve> SRS<G> {
         }
 
         // verify the equation
-        let scalars: Vec<_> = scalars.iter().map(|x| x.into_repr()).collect();
-        VariableBaseMSM::multi_scalar_mul(&points, &scalars) == G::Projective::zero()
+        G::Group::msm_unchecked(&points, scalars.as_slice()) == G::Group::zero()
```
Comment (mimoo) [inline poly-commitment/src/commitment.rs:869]: I'm wondering if this should be a checked `msm` here

File: poly-commitment/src/commitment.rs [hunk 5]
```diff
@@ -1089,20 +1091,20 @@ pub mod caml {
...
-        G: AffineCurve,
+        G: AffineRepr,
         CamlG: From<G>,
     {
         fn from(polycomm: PolyComm<G>) -> Self {
             Self {
-                unshifted: polycomm.unshifted.into_iter().map(Into::into).collect(),
-                shifted: polycomm.shifted.map(Into::into),
+                unshifted: polycomm.unshifted.into_iter().map(From::from).collect(),
+                shifted: polycomm.shifted.map(From::from),
```
Comment (mimoo) [inline poly-commitment/src/commitment.rs:1100]: this change is intriguing

File: poly-commitment/src/commitment.rs [hunk 6]
```diff
@@ -1160,19 +1162,19 @@ pub mod caml {
...
+                    .map(|(g1, g2)| (From::from(g1), From::from(g2)))
                     .collect(),
-                delta: opening_proof.delta.into(),
-                z1: opening_proof.z1.into(),
-                z2: opening_proof.z2.into(),
-                sg: opening_proof.sg.into(),
+                delta: From::from(opening_proof.delta),
+                z1: From::from(opening_proof.z1),
+                z2: From::from(opening_proof.z2),
+                sg: From::from(opening_proof.sg),
```
Comment (mimoo) [inline poly-commitment/src/commitment.rs:1170]: why all of these changes?

File: poly-commitment/src/srs.rs [hunk 7]
```diff
@@ -164,42 +164,47 @@ where
...
             let num_terms = min((i + 1) * srs_size, n) - start_offset;
             for j in 0..num_terms {
-                lg[start_offset + j] = self.g[j].into_projective()
+                lg[start_offset + j] = self.g[j].into_group()
             }
             // Apply the IFFT
             domain.ifft_in_place(&mut lg);
-            <G as AffineCurve>::Projective::batch_normalization(lg.as_mut_slice());
             // Append the 'partial Langrange polynomials' to the vector of unshifted chunks
+            let lg = G::Group::normalize_batch(lg.as_mut_slice())
```
Comment (mimoo) [inline poly-commitment/src/srs.rs:177]:
```
            let lg = G::Group::normalize_batch(&lg)
```

**Review by mimoo — APPROVED** [review]
This is amazing! Thank-you so much for the PR :) and apologize for taking so long to review it. I'm going to start working on the Mina side

File: poly-commitment/src/commitment.rs (hunk 4 above)
Comment (mimoo) [reply poly-commitment/src/commitment.rs:869]: https://github.com/o1-labs/proof-systems/pull/979 actually fixes the code so that it's always a checked `msm`, so we don't have to care too much in this PR and changing everything to `msm_unchecked` is good enough at this point

File: curves/src/pasta/curves/mod.rs [hunk 8]
```diff
@@ -1,5 +1,8 @@
 pub mod pallas;
 pub mod vesta;
 
+pub use pallas::{Affine as Pallas, PallasConfig as PallasParameters, PallasConfig};
+pub use vesta::{Affine as Vesta, VestaConfig as VestaParameters, VestaConfig};
```
[the remark this answers is not in the document]
Comment (imikushin) [reply curves/src/pasta/curves/mod.rs:5]: Let me look if I can get rid of it. This does look like a temporary measure.

File: curves/src/pasta/fields/tests.rs [hunk 9]
```diff
@@ -1,24 +1,6 @@
...
-fn test_fq() {
-    let mut rng = test_rng();
-    let a: Fq = rng.gen();
-    let b: Fq = rng.gen();
-    field_test(a, b);
-    sqrt_field_test(a);
-    primefield_test::<Fq>();
-}
+test_field!(fpp; Fpp; mont_prime_field);
+test_field!(fq; Fq; mont_prime_field);
```
[the remark this answers is not in the document]
Comment (imikushin) [reply curves/src/pasta/fields/tests.rs:6]: I'd actually keep it for now: Pallas and Vesta here are not completely identical copies of the curves from arkworks (even if the difference seems tiny).

[the remark this answers is not in the document]
Comment (imikushin) [reply curves/src/pasta/fields/tests.rs:6]: Actually, these fields _are_ completely identical, you're right! Will remove.

File: curves/src/pasta/curves/pallas.rs [hunk 10]
```diff
@@ -1,70 +1,67 @@
...
+    /// COFACTOR = 1
+    const COFACTOR: &'static [u64] = &[0x1];
+
+    /// COFACTOR_INV = 1
+    const COFACTOR_INV: Fq = Fq::ONE;
 }
 
-pub type Pallas = GroupAffine<PallasParameters>;
-pub type ProjectivePallas = GroupProjective<PallasParameters>;
+pub type Affine = sw::Affine<PallasConfig>;
```
[the remark this answers is not in the document]
Comment (imikushin) [reply curves/src/pasta/curves/pallas.rs:23]: I think Pallas and Vesta implementations should actually be as close to identical as possible. Perhaps, even parameterized over base and scalar fields and the generators. IMO, this would make for a nice follow-up issue.

File: poly-commitment/src/commitment.rs (hunk 4 above)
Comment (imikushin) [reply poly-commitment/src/commitment.rs:869]: I thought about it, and decided to keep close to the original semantics which was unchecked.

File: poly-commitment/src/commitment.rs (hunk 5 above)
Comment (imikushin) [reply poly-commitment/src/commitment.rs:1100]: Yeah, Rust compiler is... demanding. These are just black magic spells which seem to please The 🦀.

Comment (imikushin) [reply poly-commitment/src/commitment.rs:1100]: I don't fully understand what's going on here. According to the docs, if `From` is implemented, Rust is supposed to give you `Into` automatically. For some reason `Into` didn't work here, so I tried `From` and it worked, so here it is 🙃

File: poly-commitment/src/commitment.rs (hunk 6 above)
Comment (imikushin) [reply poly-commitment/src/commitment.rs:1170]: You're right! These aren't necessary. Removing

File: poly-commitment/src/commitment.rs (hunk 5 above)
Comment (imikushin) [reply poly-commitment/src/commitment.rs:1100]: Here's the exact error message if these `From::from` are changed back to `Into::into`:

```
error[E0277]: a value of type `Vec<CamlG>` cannot be built from an iterator over elements of type `<G as ark_ec::AffineRepr>::Group`
    --> poly-commitment/src/commitment.rs:1094:28
     |
1094 |                 unshifted: polycomm.unshifted.into_iter().map(Into::into).collect(),
     |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ------- required by a bound introduced by this call
     |                            |
     |                            value of type `Vec<CamlG>` cannot be built from `std::iter::Iterator<Item=<G as ark_ec::AffineRepr>::Group>`
     |
     = help: the trait `FromIterator<<G as ark_ec::AffineRepr>::Group>` is not implemented for `Vec<CamlG>`
note: required by a bound in `std::iter::Iterator::collect`
    --> /Users/ivan/.rustup/toolchains/1.67.0-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:1832:19
     |
1832 |     fn collect<B: FromIterator<Self::Item>>(self) -> B
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `std::iter::Iterator::collect`
help: consider extending the `where` clause, but there might be an alternative better way to express this requirement
     |
1090 |         CamlG: From<G>, Vec<CamlG>: FromIterator<<G as ark_ec::AffineRepr>::Group>
     |                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: poly-commitment/src/commitment.rs (hunk 6 above)
Comment (imikushin) [reply poly-commitment/src/commitment.rs:1170]: Never mind: an ugly compiler error if these are reverted. Exercise for the reader.

File: poly-commitment/src/combine.rs (hunk 1 above)
Comment (imikushin) [reply poly-commitment/src/combine.rs]: Done.

File: poly-commitment/src/commitment.rs (hunk 2 above)
Comment (imikushin) [reply poly-commitment/src/commitment.rs]: Done.

File: curves/src/pasta/fields/tests.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (imikushin) [reply curves/src/pasta/fields/tests.rs]: Done.

File: curves/src/pasta/curves/pallas.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (imikushin) [reply curves/src/pasta/curves/pallas.rs]: Check out what it looks like now 😅

File: curves/src/pasta/curves/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (imikushin) [reply curves/src/pasta/curves/mod.rs]: Done. Naming is consistent now.

**Comment by mimoo** [comment]
https://github.com/MinaProtocol/mina/pull/12717 contains the change for the ocaml bindings, but I can't compile wasm stuff on my laptop anymore. cc @fabrizio-m @jspada @querolita can you guys take over and finish the wasm stuff :o ?

**Comment by imikushin** [comment]
CI is failing with a crazy phantom compile error:
```
error[E0432]: unresolved import `mina_curves::pasta::PallasParameters`
  --> kimchi/src/tests/serde.rs:35:9
   |
   |     use mina_curves::pasta::PallasParameters;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `PallasParameters` in `pasta`
```

If we try to find the string `use mina_curves::pasta::PallasParameters`, we wouldn't be able to find it anywhere in the code any more. @mimoo Any ideas what could be going on?

**Comment by mimoo** [comment]
@imikushin you might have to merge master? I'll take a look at it once I manage to make the Mina side pass CI

**Comment by imikushin** [comment]
@mimoo Merging master didn't help: same error.

**Comment by imikushin** [comment]
Hmm, merging master definitely does help locating the errors, @mimoo: they're coming from merging master 😅 Apparently something like that happens in CI, and that's a good thing!

File: rust-toolchain.toml [hunk 11]
```diff
@@ -1,4 +1,4 @@
 [toolchain]
-channel = "1.67.0"
+channel = "1.67.1"
```
Comment (mimoo) [inline rust-toolchain.toml:2]: can you revert this change actually 😅 it makes upgrades on the Mina side quite annoying

File: circuit-construction/Cargo.toml [hunk 12]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "circuit-construction"
-version = "0.1.0"
+version = "0.2.0"
```
Comment (mimoo) [inline circuit-construction/Cargo.toml:3]: let's have these version changes in a different PR as well : o

**Comment by mimoo** [comment]
Screenshot 2023-02-25 at 9 09 59 AM

so I'm currently observing these hex/decimal change in the OCaml code generated by the `expr.rs` file. It might be the move from `into_repr` to `to_bigint` but I'm not sure. Will resume investigation later

File: kimchi/src/circuits/expr.rs [hunk 13]
```diff
@@ -2343,7 +2343,7 @@ where
             JointCombiner => "joint_combiner".to_string(),
             EndoCoefficient => "endo_coefficient".to_string(),
             Mds { row, col } => format!("mds({row}, {col})"),
-            Literal(x) => format!("field(\"0x{}\")", x.into_repr()),
+            Literal(x) => format!("field(\"0x{}\")", x.into_bigint()),
```
Comment (mimoo) [inline kimchi/src/circuits/expr.rs:2346]: yeah looks like that's the change that's breaking things, I guess we need a to_hex function in `field_helpers.rs`

Comment (mimoo) [reply kimchi/src/circuits/expr.rs:2346]: oh looks like we already have one : o

**Comment by mimoo** [comment]
should be fixed here: https://github.com/o1-labs/proof-systems/pull/1027

File: circuit-construction/Cargo.toml (hunk 12 above)
Comment (mimoo) [reply circuit-construction/Cargo.toml:3]: did it in https://github.com/o1-labs/proof-systems/pull/1027

File: rust-toolchain.toml (hunk 11 above)
Comment (mimoo) [reply rust-toolchain.toml:2]: done in https://github.com/o1-labs/proof-systems/pull/1027

Comment (imikushin) [reply rust-toolchain.toml:2]: Sure thing!

**Comment by imikushin** [comment]
@mimoo I can add commits from #1027 here.
